Add integer sequences to the new BaseLanguage extensions

alamqadem opened the pull request:
## Name
Integer sequence `[0..n]`

## Operation Type

Expression

## Use Case

`[0 .. n]` would produce a sequence of integers containing all integers from `0` to `n`.
These are the 3 types of integer sequences that are supported:
- `[0..n]` finite with a fixed upper limit;
- `[0..#e]` finite with an expression as an upper limit;
- `[0..]` infinite syntax.

## Comments

dalbrecht1 committed:
included ElvisOperator runtime

alamqadem committed:
Add IntegerSequence helper

alamqadem committed:
Add IntegerSequence helper

alamqadem committed:
Add integer sequence expression

alamqadem committed:
Return empty sequence on invalid range

- add checking rules for infinite and invalid range
- update test

alamqadem committed:
Add generator for integer sequence

alamqadem committed:
Remove unused dependencies

alamqadem committed:
Delete duplicated classes

alamqadem committed:
Fix left paren editor

dalbrecht1 committed:
executed migration

dalbrecht1 commented:
1. I think compared to how the sub-string operation on strings is displayed, your sort-of implicit infinite end for the sequence is hard to spot and to fill in:

<img width="132" height="70" alt="Screenshot 2025-09-24 185316" src="https://github.com/user-attachments/assets/6e68eacb-25df-4d6d-bdf9-3e97bea1da16" />

There should be some field (or visual gap) that the user at least would know to try to put something in, so I guess if the "end"-position for the sequence would be a grayed out "inf" by default (and with proper/expected spacing between the cells), that would feel more natural.

2. I tend to interpret warnings that I receive from the compiler as "well, that's certainly not ideal, the way you wrote it, and no-one will guarantee you that this works forever, but _technically_ it's fine; but I would really appreciate it you found another way of writing it." In contrast, while an endless integer sequence is certainly potentially dangerous if evaluated in full, there is no other way to write it, so, I'm not sure I get why writing one should result in a warning. But that might be a philosophical question. Furthermore, given that your default state is the infinite end, having the element appear with a warning directly is certainly also not ideal...

3. I added another commit on top, because MPS asked me to migrate the code. I'm not exactly sure why, but at least some resolve-info's were corrected this way.

4. We should really adopt a better separation of the individual features in the modules using virtual packages. So, great that you started that here! I'll (have to) move my declarations in my next PR.

5. For my own learning effect: Was there a clear benefit of explicitly modeling the Range Bounds, instead of just having one mandatory "Expression" for the start, and one optional for the end? Restricted by a typing rule. To be honest, I still don't fully get when you should just use general types (like Expression) and when the whole "# real-code-here" style is used/preferred.

6. There is a lot going on on that branch, so accepting that some details of the history will get lost, I think we should squash that, once it is ready.

alamqadem committed:
show inf placeholder in editor when right bound is not specified

alamqadem committed:
change warning to info in check for infinite range

alamqadem commented:
Thank you for the review.

> 1. I think compared to how the sub-string operation on strings is displayed, your sort-of implicit infinite end for the sequence is hard to spot and to fill in:
> 
> There should be some field (or visual gap) that the user at least would know to try to put something in, so I guess if the "end"-position for the sequence would be a grayed out "inf" by default (and with proper/expected spacing between the cells), that would feel more natural.

I've add `inf` when the right bound is missing. Do you think it's better now?
<img width="563" height="103" alt="image" src="https://github.com/user-attachments/assets/e43073bf-9e44-4323-91c9-67c687e5cee6" />


> 2. I tend to interpret warnings that I receive from the compiler as "well, that's certainly not ideal, the way you wrote it, and no-one will guarantee you that this works forever, but _technically_ it's fine; but I would really appreciate it you found another way of writing it." In contrast, while an endless integer sequence is certainly potentially dangerous if evaluated in full, there is no other way to write it, so, I'm not sure I get why writing one should result in a warning. But that might be a philosophical question. Furthermore, given that your default state is the infinite end, having the element appear with a warning directly is certainly also not ideal...

I've changed the warning to an info instead. What do you think?

> 3. I added another commit on top, because MPS asked me to migrate the code. I'm not exactly sure why, but at least some resolve-info's were corrected this way.

Thanks looking at the changes it looks like only metadata changes.

> 5. For my own learning effect: Was there a clear benefit of explicitly modeling the Range Bounds, instead of just having one mandatory "Expression" for the start, and one optional for the end? Restricted by a typing rule. To be honest, I still don't fully get when you should just use general types (like Expression) and when the whole "# real-code-here" style is used/preferred.

Here I wanted to have something like the light quotation node, just because people are familiar with that syntax. I allow numbers only and expressions only if the `#` is used.

> 6. There is a lot going on on that branch, so accepting that some details of the history will get lost, I think we should squash that, once it is ready.

Good point. I will squash merge the PR once it's approved.

dalbrecht1 committed:
minor fix for infinite semantic

dalbrecht1 commented:
1. Yes, that's soo much cleaner and more intuitive!
2. I mean, meanwhile I saw that "while(true)" also gets flagged by the MPS type checker, this is almost the same, but also strange to me. But yes, having less force towards one specific style of coding is always appreciated.

<To have the the enumeration restart...>

5. Understood. Thanks for the explanation/clarification.
6. I'll just approve and merge in a second.
7. Another new point, I just found out that the infinite sequence will always start at "0" despite any other given number in the editor. I just fixed that myself, in order to finally get this PR closed.
